tests/oeedger8r/enc/testdeepcopy.cpp

```diff
@@ -0,0 +1,136 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+#include "../edltestutils.h"
+
+#include <openenclave/enclave.h>
+#include <openenclave/internal/tests.h>
+#include "all_t.h"
+
+static uint64_t data[8] = {0x1112131415161718,
+                           0x2122232425262728,
+                           0x3132333435363738,
+                           0x4142434445464748,
+                           0x5152535455565758,
+                           0x6162636465666768,
+                           0x7172737475767778,
+                           0x8182838485868788};
+
+// Assert that the struct is copied by value, such that `s.ptr` is the
+// address of `data[]` in the host (also passed via `ptr`).
+void deepcopy_value(ShallowStruct s, uint64_t* ptr)
+{
+    OE_TEST(s.count == 7);
+    OE_TEST(s.size == 64);
+    OE_TEST(s.ptr == ptr);
+    OE_TEST(oe_is_outside_enclave(s.ptr, sizeof(uint64_t)));
+}
+
+// Assert that the struct is shallow-copied (even though it is passed
+// by pointer), such that `s->ptr` is the address of `data[]` in the
+// host (also passed via `ptr`).
+void deepcopy_shallow(ShallowStruct* s, uint64_t* ptr)
+{
+    OE_TEST(s->count == 7);
+    OE_TEST(s->size == 64);
+    OE_TEST(s->ptr == ptr);
+    OE_TEST(oe_is_outside_enclave(s->ptr, sizeof(uint64_t)));
+}
+
+// Assert that the struct is deep-copied such that `s->ptr` has a copy
+// of three elements of `data` in enclave memory.
+void deepcopy_count(CountStruct* s)
+{
+    OE_TEST(s->count == 7);
```

Comment:
Do these tests (and fields) here add any value? I was about to suggest that the oe_is_within_enclave ought to use size or count and then realized that those fields weren't being used as annotations.

Reply:
They're just being really explicit based on the documented scenarios (which also state what each value should be before and after).

Reply:
Ok.